homeassistant/components/tplink/entity.py

```diff
@@ -57,24 +78,201 @@ async def _async_wrap(self: _T, *args: _P.args, **kwargs: _P.kwargs) -> None:
+    parent: Device | None = None,
+) -> list[_E]:
+    """Return a list of entities to add.
+
+    This filters out unwanted features to avoid creating unnecessary entities
+    for device features that are implemented by specialized platforms like light.
+    """
+
+    def _filter(dev: Device, feat: Feature) -> bool:
+        if feat.type != feature_type:
+            return False
+
+        # We skip primary features for device types that have specialized platforms.
+        ignore_primary_controls_devicetypes = [
+            DeviceType.Bulb,
+            DeviceType.LightStrip,
+            DeviceType.Dimmer,
+            DeviceType.Thermostat,
+        ]
+        if (
+            feat.category == Feature.Category.Primary
+            and dev.device_type in ignore_primary_controls_devicetypes
+        ):
+            return False
+
+        return True
+
+    return [
+        entity_class(device, coordinator, feat, parent=parent)
+        for feat in device.features.values()
+        if _filter(device, feat)
+    ]
+
+
+def _entities_for_device_and_its_children[_E: CoordinatedTPLinkEntity](
+    device: Device,
+    *,
+    feature_type: Feature.Type,
+    entity_class: type[_E],
+    coordinator: TPLinkDataUpdateCoordinator,
+) -> list[_E]:
+    """Create entities for device and its children.
+
+    This is a helper that calls *_entities_for_device* for the device and its children.
+    """
+    entities = []
+    if device.children:
+        _LOGGER.debug("Initializing device with %s children", len(device.children))
+        for child in device.children:
+            entities.extend(
+                _entities_for_device(
+                    child,
+                    feature_type=feature_type,
+                    entity_class=entity_class,
+                    coordinator=coordinator,
+                    parent=device,
+                )
+            )
+
+    entities.extend(
+        _entities_for_device(
+            device,
+            feature_type=feature_type,
+            entity_class=entity_class,
+            coordinator=coordinator,
+        )
+    )
+
+    return entities
+
+
+def _description_for_feature[_D: EntityDescription](
+    desc_cls: type[_D], feature: Feature, **kwargs: Unpack[EntityDescriptionExtras]
+) -> _D:
+    """Return description object for the given feature.
+
+    This is responsible for setting the common parameters & deciding based on feature id
+    which additional parameters are passed.
+    """
+
+    # Disable all debug features that are not explicitly enabled.
+    if "entity_registry_enabled_default" not in kwargs:
+        kwargs["entity_registry_enabled_default"] = (
+            feature.category is not Feature.Category.Debug
+        )
+
+    return desc_cls(
+        key=feature.id,
+        translation_key=feature.id,
+        name=feature.name,
+        # Setting an icon overrides the translation, so we doing it here.
+        # icon=feature.icon,
```

Comment:
Homeassistant currently picks the `icon` over an icon translation. It would be great to have it the other way around, i.e., to show a relevant icon if there's no translation available, but I couldn't find a way to check for the existence of a translation.

Reply:
Removed this for now. The proper way is for us to dump a list of available features & add them into the icon translations file.